File: code/scenes/src/main/java/org/fbme/scenes/controllers/components/ComponentsExtensionsFacility.java
package org.fbme.scenes.controllers.components;

import jetbrains.mps.nodeEditor.cells.EditorCell;
import org.fbme.scenes.controllers.scene.*;
import org.fbme.scenes.viewmodel.ComponentExtsView;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.awt.*;
import java.util.List;
import java.util.*;

public class ComponentsExtensionsFacility<CompT, CompExtT, FormT> {

    private final Map<CompExtT, ComponentExtEntry<CompT, CompExtT, FormT>> myComponentExts = new HashMap<>();

    private final ComponentExtsView<CompT, CompExtT> myExtView;
    final ComponentsFacility<CompT, FormT> myComponentsFacility;
    final ComponentExtControllerFactory<CompExtT, FormT> myControllerFactory;

    public ComponentsExtensionsFacility(ComponentExtsView<CompT, CompExtT> extView, ComponentsFacility<CompT, FormT> componentsFacility, ComponentExtControllerFactory<CompExtT, FormT> controllerFactory, Layer compExtsLayer, Layer tracesLayer) {
        this.myExtView = extView;
        this.myComponentsFacility = componentsFacility;
        this.myControllerFactory = controllerFactory;

        init();

        SceneEditor editor = componentsFacility.myEditor;
        editor.addLayouter(new MyLayouter());
        editor.addCellProvider(compExtsLayer, new MyCellProvider());
        editor.addPainter(tracesLayer, new MyTracesPainter());
    }

    private void init() {
        Set<CompT> viewComponents = myComponentsFacility.myView.getComponents();
        Map<CompExtT, CompT> viewComponentExts = new HashMap<>();
        for (CompT viewComponent : viewComponents) {
            for (CompExtT viewComponentExt : myExtView.getExtensions(viewComponent)) {
                viewComponentExts.put(viewComponentExt, viewComponent);
            }
        }

        for (Map.Entry<CompExtT, CompT> component : viewComponentExts.entrySet()) {
            ComponentExtEntry<CompT, CompExtT, FormT> entry = new ComponentExtEntry<>(this, component.getKey(), component.getValue());
            myComponentExts.put(component.getKey(), entry);
        }
    }

    @NotNull
    public FormT getModelForm(@NotNull CompExtT componentExt) {
        return myComponentExts.get(componentExt).myComponentEntry.myModelForm;
    }

    @Nullable
    public FormT getTransformedForm(@NotNull CompExtT componentExt) {
        return myComponentExts.get(componentExt).myComponentEntry.myTransformedForm;
    }

    public final ComponentExtController<FormT> getController(@NotNull CompExtT extension) {
        return myComponentExts.get(extension).myController;
    }

    private class MyCellProvider implements EditorCellProvider {

        @Override
        public List<EditorCell> getCells() {
            ArrayList<EditorCell> ret = new ArrayList<>();
            for (ComponentExtEntry<CompT, CompExtT, FormT> componentExt : myComponentExts.values()) {
                ret.add(componentExt.myController.getComponentCell());
            }
            return ret;
        }
    }

    private class MyLayouter implements SceneLayouter {

        @Override
        public void relayout() {
            for (ComponentExtEntry<CompT, CompExtT, FormT> componentExt : myComponentExts.values()) {
                ComponentEntry<CompT, FormT> componentEntry = componentExt.myComponentEntry;
                componentExt.myController.updateCellWithForm((componentEntry.myTransformedForm != null ? componentEntry.myTransformedForm : componentEntry.myModelForm));
            }
        }

        @NotNull
        @Override
        public Rectangle getLayoutBounds() {
            return new Rectangle(-1, -1);
        }
    }

    private class MyTracesPainter implements ScenePainter {

        @Override
        public void paint(Graphics2D graphics) {
            for (ComponentExtEntry<CompT, CompExtT, FormT> componentExt : myComponentExts.values()) {
                ComponentExtController<FormT> controller = componentExt.myController;
                ComponentEntry<CompT, FormT> componentEntry = myComponentsFacility.myComponents.get(componentExt.myComponent);
                FormT form = (componentEntry.myTransformedForm != null ? componentEntry.myTransformedForm : componentEntry.myModelForm);
                controller.paintTrace(graphics.create(), form);
            }
        }
    }
}

File: code/scenes/src/main/kotlin/org/fbme/scenes/controllers/components/ComponentsExtensionsFacility.kt
package org.fbme.scenes.controllers.components

import jetbrains.mps.nodeEditor.cells.EditorCell
import org.fbme.scenes.controllers.scene.EditorCellProvider
import org.fbme.scenes.controllers.scene.Layer
import org.fbme.scenes.controllers.scene.SceneLayouter
import org.fbme.scenes.controllers.scene.ScenePainter
import org.fbme.scenes.viewmodel.ComponentExtensionsView
import java.awt.Graphics2D
import java.awt.Rectangle

class ComponentsExtensionsFacility<CompT, CompExtT, FormT>(
    extensionsView: ComponentExtensionsView<CompT, CompExtT>,
    componentsFacility: ComponentsFacility<CompT, FormT>,
    controllerFactory: ComponentExtControllerFactory<CompExtT, FormT>,
    componentExtensionsLayer: Layer,
    tracesLayer: Layer
) {
    private val componentExtensions: MutableMap<CompExtT, ComponentExtEntry<CompT, CompExtT, FormT>> = HashMap()
    private val extView: ComponentExtensionsView<CompT, CompExtT>
    val componentsFacility: ComponentsFacility<CompT, FormT>
    val controllerFactory: ComponentExtControllerFactory<CompExtT, FormT>
    private fun init() {
        val viewComponents = componentsFacility.view.components
        val viewComponentExtensions: MutableMap<CompExtT, CompT> = HashMap()
        for (viewComponent in viewComponents) {
            for (viewComponentExt in extView.getExtensions(viewComponent!!)) {
                viewComponentExtensions[viewComponentExt] = viewComponent
            }
        }
        for ((key, value) in viewComponentExtensions) {
            val entry = ComponentExtEntry(this, key, value)
            componentExtensions[key] = entry
        }
    }

    fun getModelForm(extension: CompExtT): FormT {
        val componentExtEntry = componentExtensions[extension] ?: error("Component extension not found")
        return componentExtEntry.componentEntry.modelForm
    }

    fun getTransformedForm(extension: CompExtT): FormT? {
        val componentExtEntry = componentExtensions[extension] ?: error("Component extension not found")
        return componentExtEntry.componentEntry.transformedForm
    }

    fun getController(extension: CompExtT): ComponentExtController<FormT> {
        val componentExtEntry = componentExtensions[extension] ?: error("Component extension not found")
        return componentExtEntry.controller
    }

    private inner class MyCellProvider : EditorCellProvider {
        override val cells: List<EditorCell>
            get() {
                val ret = ArrayList<EditorCell>()
                for (componentExt in componentExtensions.values) {
                    val componentExtController = componentExt.controller
                    ret.add(componentExtController.componentCell)
                }
                return ret
            }
    }

    private inner class MyLayouter : SceneLayouter {
        override fun relayout() {
            for (componentExt in componentExtensions.values) {
                val componentEntry = componentExt.componentEntry
                componentExt.controller.updateCellWithForm(componentEntry.transformedForm ?: componentEntry.modelForm)
            }
        }

        override val layoutBounds: Rectangle
            get() = Rectangle(-1, -1)
    }

    private inner class MyTracesPainter : ScenePainter {
        override fun paint(graphics: Graphics2D) {
            for (componentExt in componentExtensions.values) {
                val controller = componentExt.controller
                val componentEntry =
                    componentsFacility.components[componentExt.component] ?: error("Component not found")
                controller.paintTrace(graphics.create(), componentEntry.transformedForm ?: componentEntry.modelForm)
            }
        }
    }

    init {
        this.extView = extensionsView
        this.componentsFacility = componentsFacility
        this.controllerFactory = controllerFactory
        init()
        val editor = componentsFacility.editor
        editor.addLayouter(MyLayouter())
        editor.addCellProvider(componentExtensionsLayer, MyCellProvider())
        editor.addPainter(tracesLayer, MyTracesPainter())
    }
}
